src/puppetlabs/puppetdb/middleware.clj

```diff
@@ -293,3 +290,23 @@
       wrap-with-certificate-cn
       wrap-with-default-body
       wrap-with-debug-logging))
+
+(defn wrap-with-parent-check
+  "Middleware that checks the parent exists before serving the rest of the
+   application. This ensures we always return 404's on child paths when the
+   parent data is empty."
+  [app version parent id]
+  (fn [{:keys [globals] :as req}]
+    (let [{:keys [scf-read-db url-prefix]} globals
+          status-fn (case parent
+                      :node nodes/node-exists?
+                      :environment envs/environment-exists?
+                      :report reports/report-exists?)]
+      ;; There is a race condition here, in particular we open up 1 transaction
+      ;; for the parent test, but the rest of the query results are done via the
+      ;; streaming query. This can't be solved until we work out a way to
+      ;; pass through an existing db handle through to the streamed query thread.
```

Comment:
Of course, its highly unlikely, and if a user tries again he will get a 404. We should really bring about a middleware that does this for us (Ryan had one, I'm removing it this time because its dead code right now, and minor anyway) and thread that db variable through to the backend thread of the query potential. But @senior and I talked about it, I think the upgrade to clojure.java.jdbc where threading the db pointer becomes mandatory would make this proposed change more clean.